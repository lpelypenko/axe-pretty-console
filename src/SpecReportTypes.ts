import { Result } from 'axe-core';

interface Summary {
    description: string;
    id: string;
    wcag: string;
    impact: string;
    nodes: number;
}

export interface SpecPreparedData {
    violationsTotal: string;
    violationsSummaryTable: Summary[] | [];
    violationsDetailsFormatted: string;
    checksPassedSummary?: string;
}

export interface SpecReportInput {
    violations: Result[];
    passes?: Result[];
    url?: string;
    skipResultTable?: boolean;
}
