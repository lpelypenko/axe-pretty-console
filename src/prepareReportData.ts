import { bold } from 'chalk';
import { SpecPreparedData, SpecReportInput } from './SpecReportTypes';

/**
 * Axe returns variety of tags that are not necessary for our purposes.
 * We are interested only in WCAG related tags and Best Practices.
 * Function tries to determine if tag belongs to Best Practice or any WCAG 2.x, otherwise all tags will be returned raw
 * @param tags
 * @returns {string}
 */
export function getWcagReference(tags: string[]): string {
    const tagsNames = ['wcag2a', 'wcag2aa', 'wcag21a', 'wcag21aa', 'best-practice'];
    const foundTags = tags.filter((tag) => tagsNames.includes(tag));

    const tagNamesToAccessibilityStandard: Record<string, string> = {
        wcag2a: 'WCAG 2.0 Level A',
        wcag2aa: 'WCAG 2.0 Level AA',
        wcag21a: 'WCAG 2.1 Level A',
        wcag21aa: 'WCAG 2.1 Level AA',
        'best-practice': 'Best practice',
    };

    if (foundTags.length > 0) {
        return foundTags.map((tag) => tagNamesToAccessibilityStandard[tag]).join(',');
    }

    return tags.join(',');
}

/**
 * Prepare report splitting it into sections:
 * - total accessibility violations (counting nodes)
 * - summary of violations that could be printed as table
 * - detailed list of violations that could be printed as formatted text
 * @param violations
 * @param passes
 * @param url
 */
export function prepareReportData({ violations, passes, url }: SpecReportInput): SpecPreparedData {
    if (!violations) {
        throw new TypeError(
            'prepareReportData() requires violations to be passed as an object: prepareReportData({violations: Result[]})'
        );
    }
    const passesIds = passes
        ? passes
              .map(({ id }) => {
                  return id;
              })
              .join(', ')
        : '';
    const checksPassedSummary = passes
        ? `${bold(`Page passed ${passes.length} axe rules:`)} ${passesIds}`
        : '';
    if (violations.length === 0) {
        return {
            violationsTotal: `No accessibility violation were detected${
                url ? ` for the ${url}` : ''
            }`,
            violationsSummaryTable: [],
            violationsDetailsFormatted: '',
            checksPassedSummary,
        };
    }
    const total = violations.reduce((acc, { nodes }) => {
        acc += nodes.length;
        return acc;
    }, 0);
    const violationsTotal = `Axe core library found ${total} violation${total === 1 ? '' : 's'}${
        url ? ` for the ${url}` : ''
    }`;
    // Prepare data to show summary
    const violationsSummaryTable = violations.map(({ nodes, description, id, tags, impact }) => ({
        description,
        id,
        wcag: getWcagReference(tags),
        impact: impact || 'n/a',
        nodes: nodes.length,
    }));

    // Prepare data to show detailed list of violations
    const violationsDetailsFormatted = violations
        .map(({ nodes, description, help, id, tags, helpUrl }, i) => {
            const wcagRef = `   ${bold('WCAG: ')}'${getWcagReference(tags)}'`;
            const axeRuleId = `${bold('id: ')}'${id}'`;
            const summary = `   ${bold('description: ')}${description}`;
            const name = `   ${bold('name: ')}${help}`;
            const learnMore = `${bold('learn more: ')}${helpUrl}`;
            const nodeDetails = `${bold('   Affected elements:')}\n`.concat(
                nodes
                    .map(({ target, html }) => {
                        const targetNodes = target.map((node) => `"${node}"`).join(', ');
                        return `\tSelector: ${targetNodes}\tSource code: ${html}`;
                    })
                    .join('\n')
            );

            return `${
                i + 1
            }. ${axeRuleId}\t${learnMore}\n${name}\n${summary}\n${wcagRef}\n${nodeDetails}\n`;
        })
        .join('\n');

    return {
        violationsTotal,
        violationsSummaryTable,
        violationsDetailsFormatted,
        checksPassedSummary,
    };
}
