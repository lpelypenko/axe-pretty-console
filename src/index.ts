import { prepareReportData } from './prepareReportData';
import { SpecReportInput } from './SpecReportTypes';

/**
 * pretty print outputs only violations and passes
 * @param report
 */
export function prettyPrintAxeReport(report: SpecReportInput): void {
    try {
        const preparedData = prepareReportData({
            violations: report.violations,
            passes: report.passes,
            url: report.url,
        });
        // Print summary of violations as a table
        if (preparedData.violationsSummaryTable.length !== 0) {
            console.info(preparedData.violationsTotal);
            if (!report.skipResultTable) {
                console.table(preparedData.violationsSummaryTable);
            }
            console.info(preparedData.violationsDetailsFormatted);
        }
        if (preparedData.checksPassedSummary !== '') {
            console.info(preparedData.checksPassedSummary);
        }
    } catch (e) {
        // throw the error only if user made a mistake in passing wrong variable
        if (e instanceof TypeError) {
            throw e;
        }
        console.log(`printAxeReport() failed with an error${e}`);
    }
}
