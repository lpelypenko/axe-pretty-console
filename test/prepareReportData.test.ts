import { bold } from 'chalk';
import { prepareReportData, getWcagReference } from '../src/prepareReportData';

const axeRawViolations = require('./__mock_data__/rawViolations.json');
const axeRawPasses = require('./__mock_data__/rawPasses.json');

describe('prepareReportData() test', () => {
    it('Throws an error if violations are not passed', async () => {
        expect(() => {
            prepareReportData(axeRawViolations);
        }).toThrow(
            'prepareReportData() requires violations to be passed as an object: prepareReportData({violations: Result[]})'
        );
    });
    it('Returns full report with violations', async () => {
        const report = prepareReportData({ violations: axeRawViolations });
        expect(report.violationsSummaryTable).toMatchSnapshot();
        expect(report.violationsDetailsFormatted).not.toEqual('');
        expect(report.violationsTotal).toEqual('Axe core library found 30 violations');
        expect(report.checksPassedSummary).toEqual('');
    });
    it('Returns full report with violations and passes', async () => {
        const report = prepareReportData({ violations: axeRawViolations, passes: axeRawPasses });
        expect(report.violationsSummaryTable).toMatchSnapshot();
        expect(report.violationsDetailsFormatted).not.toEqual('');
        expect(report.violationsTotal).toEqual('Axe core library found 30 violations');
        expect(report.checksPassedSummary).toEqual(`${bold('Page passed 6 axe rules:')} page-has-heading-one, region, tabindex, table-duplicate-name, table-fake-caption, td-headers-attr`);
    });
    it('Verify returns empty report', async () => {
        expect(prepareReportData({ violations: [] })).toEqual({
            checksPassedSummary: '',
            violationsDetailsFormatted: '',
            violationsSummaryTable: [],
            violationsTotal: 'No accessibility violation were detected',
        });
    });
});

describe('getWcagReference() test with different tags as an input', () => {
    it.each([
        [
            ['cat.name-role-value', 'wcag2a', 'wcag412', 'section508', 'section508.22.a'],
            'WCAG 2.0 Level A',
        ],
        [['cat.keyboard', 'best-practice'], 'Best practice'],
        [['cat.color', 'wcag2aa', 'wcag143'], 'WCAG 2.0 Level AA'],
        [['cat.color', 'experimental'], 'cat.color,experimental'],
    ])(
        'Verify wcag tag determined correctly with getWcagReference(%o) expecting (%s) in return',
        async (input: string[], expectedOutput: string) => {
            expect(await getWcagReference(input)).toEqual(expectedOutput);
        }
    );
});
