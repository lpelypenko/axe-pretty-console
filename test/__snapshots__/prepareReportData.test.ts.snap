// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareReportData() test Returns full report with violations 1`] = `
Array [
  Object {
    "description": "Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds",
    "id": "color-contrast",
    "impact": "serious",
    "nodes": 9,
    "wcag": "WCAG 2.0 Level AA",
  },
  Object {
    "description": "Ensures every id attribute value is unique",
    "id": "duplicate-id",
    "impact": "minor",
    "nodes": 1,
    "wcag": "WCAG 2.0 Level A",
  },
  Object {
    "description": "Ensures every form element has a label",
    "id": "label",
    "impact": "critical",
    "nodes": 6,
    "wcag": "WCAG 2.0 Level A",
  },
  Object {
    "description": "Ensures the document has a main landmark",
    "id": "landmark-one-main",
    "impact": "moderate",
    "nodes": 1,
    "wcag": "Best practice",
  },
  Object {
    "description": "Ensures all page content is contained by landmarks",
    "id": "region",
    "impact": "moderate",
    "nodes": 8,
    "wcag": "Best practice",
  },
  Object {
    "description": "Ensures svg elements with an img, graphics-document or graphics-symbol role have an accessible text",
    "id": "svg-img-alt",
    "impact": "serious",
    "nodes": 5,
    "wcag": "WCAG 2.0 Level A",
  },
]
`;

exports[`prepareReportData() test Returns full report with violations and passes 1`] = `
Array [
  Object {
    "description": "Ensures the contrast between foreground and background colors meets WCAG 2 AA contrast ratio thresholds",
    "id": "color-contrast",
    "impact": "serious",
    "nodes": 9,
    "wcag": "WCAG 2.0 Level AA",
  },
  Object {
    "description": "Ensures every id attribute value is unique",
    "id": "duplicate-id",
    "impact": "minor",
    "nodes": 1,
    "wcag": "WCAG 2.0 Level A",
  },
  Object {
    "description": "Ensures every form element has a label",
    "id": "label",
    "impact": "critical",
    "nodes": 6,
    "wcag": "WCAG 2.0 Level A",
  },
  Object {
    "description": "Ensures the document has a main landmark",
    "id": "landmark-one-main",
    "impact": "moderate",
    "nodes": 1,
    "wcag": "Best practice",
  },
  Object {
    "description": "Ensures all page content is contained by landmarks",
    "id": "region",
    "impact": "moderate",
    "nodes": 8,
    "wcag": "Best practice",
  },
  Object {
    "description": "Ensures svg elements with an img, graphics-document or graphics-symbol role have an accessible text",
    "id": "svg-img-alt",
    "impact": "serious",
    "nodes": 5,
    "wcag": "WCAG 2.0 Level A",
  },
]
`;
